### PR TITLE
create a repo management tool for DevTools (devman)

### PR DESCRIPTION
**Pull request by devoncarew**
- create a repo management tool for DevTools

Here's the UX currently:

```
A repo management tool for DevTools.

Usage: devman <command> [arguments]

Global options:
-h, --help    Print this usage information.

Available commands:
  analyze        Analyze all DevTools packages.
  help           Display help information for devman.
  list           List all the DevTools packages.
  packages-get   Run 'pub get' in all DevTools packages.

Run "devman help <command>" for more information about a command.
```

It's run via `dart ./tool/bin/devman.dart`, but we should add a shell script to simplify that.

Feedback on the cli tool name is appreciated (`devman` currently). I didn't want to use something that would interfere with shell command completion for `devtools`.

It adds two main commands - `packages-get` and `analyze`. Once this PR is reviewed and lands, we can update the bots to use these commands. That'll ensure that the tool continues to work.

### CODE REVIEW

File: tool/bin/devman.dart [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+void main(List<String> args) async {
+  final DevToolsCommandRunner runner = DevToolsCommandRunner();
+  try {
+    final dynamic result = await runner.run(args);
+    exit(result is int ? result : 0);
+  } catch (e) {
+    if (e is UsageException) {
+      stderr.writeln('$e');
+      exit(64);
```
Comment (kenzieschmoll) [inline tool/bin/devman.dart]: what is 64? Perhaps put in a const and explain what this exit code is

File: tool/bin/devman.dart [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    } else {
+      stderr.writeln('$e');
+      exit(1);
```
Comment (kenzieschmoll) [inline tool/bin/devman.dart]: same with 1 for consistency

File: tool/lib/commands/analyze.dart [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+class AnalyzeCommand extends Command {
+  @override
+  String get name => 'analyze';
+
+  @override
+  String get description => 'Analyze all DevTools packages.';
+
+  @override
+  Future run() async {
+    final FlutterSdk sdk = FlutterSdk.getSdk();
```
Comment (kenzieschmoll) [inline tool/lib/commands/analyze.dart]: remove type

Here and elsewhere in this method

File: tool/lib/commands/analyze.dart [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    for (Package p in packages) {
+      if (!p.hasAnyDartCode) {
+        continue;
+      }
+
+      final Progress progress = log.progress('  ${p.relativePath}');
+
+      final Process process = await Process.start(
+          sdk.flutterToolPath, ['--no-color', 'analyze'],
+          workingDirectory: p.packagePath);
```
Comment (kenzieschmoll) [inline tool/lib/commands/analyze.dart]: nit: trailing comma

File: tool/lib/commands/list.dart [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+class ListCommand extends Command {
+  @override
+  String get name => 'list';
+
+  @override
+  String get description => 'List all the DevTools packages.';
+
+  @override
+  Future run() async {
+    final DevToolsRepo repo = DevToolsRepo.getInstance();
```
Comment (kenzieschmoll) [inline tool/lib/commands/list.dart]: same: remove type

File: tool/lib/commands/packages_get.dart [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+import '../model.dart';
+import '../utils.dart';
+
+class PackagesGetCommand extends Command {
+  @override
+  String get name => 'packages-get';
+
+  @override
+  String get description => "Run 'pub get' in all DevTools packages.";
```
Comment (kenzieschmoll) [inline tool/lib/commands/packages_get.dart]: nit: r/pub get/flutter pub get/

File: tool/lib/commands/packages_get.dart [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      return 1;
+    }
+
+    final Logger log = Logger.standard();
+
+    final DevToolsRepo repo = DevToolsRepo.getInstance();
+
+    final List<Package> packages = repo.getPackages();
+
+    log.stdout('Running flutter packages get...');
```
Comment (kenzieschmoll) [inline tool/lib/commands/packages_get.dart]: r/flutter packages get/flutter pub get

When you run `flutter packages get` or `flutter pub get`, flutter tool prints `Running "flutter pub get" in devtools_app...  `

File: tool/lib/commands/packages_get.dart [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    int failureCount = 0;
+
+    for (Package p in packages) {
+      final Progress progress = log.progress('  ${p.relativePath}');
+
+      final Process process = await Process.start(
+          sdk.flutterToolPath, ['packages', 'get'],
+          workingDirectory: p.packagePath);
```
Comment (kenzieschmoll) [inline tool/lib/commands/packages_get.dart]: trailing comma

File: tool/lib/model.dart [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  @override
+  String toString() => '[DevTools $repoPath]';
+
+  /// This returns the DevToolsRepo instance based on the current working
+  /// directory.
+  ///
+  /// This can fail and return null if the current working directory is not
+  /// contained within a git checkout of DevTools.
+  static DevToolsRepo getInstance() {
+    final String repoPath = _findRepoRoot(Directory.current);
```
Comment (kenzieschmoll) [inline tool/lib/model.dart]: remove types throughout this file as well

File: tool/lib/model.dart [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    // Look for it relative to the current Dart process.
+    final String dartVmPath = Platform.resolvedExecutable;
+    final List<String> pathSegments = path.split(dartVmPath);
+    final List<String> expectedSegments =
+        path.posix.split('bin/cache/dart-sdk/bin/dart');
+
+    if (pathSegments.length >= expectedSegments.length) {
+      // Remove the trailing 'dart'.
+      pathSegments.removeLast();
+      expectedSegments.removeLast();
```
Comment (kenzieschmoll) [inline tool/lib/model.dart]: why is it necessary for this to be done outside of the while loop

File: tool/lib/commands/analyze.dart (hunk 3 above)
Comment (devoncarew) [reply tool/lib/commands/analyze.dart]: done, and removed in several more places in this PR

File: tool/lib/commands/analyze.dart (hunk 4 above)
Comment (devoncarew) [reply tool/lib/commands/analyze.dart]: done

File: tool/lib/commands/list.dart (hunk 5 above)
Comment (devoncarew) [reply tool/lib/commands/list.dart]: done

File: tool/lib/commands/packages_get.dart (hunk 6 above)
Comment (devoncarew) [reply tool/lib/commands/packages_get.dart]: done

File: tool/lib/commands/packages_get.dart (hunk 7 above)
Comment (devoncarew) [reply tool/lib/commands/packages_get.dart]: updated

File: tool/lib/commands/packages_get.dart (hunk 8 above)
Comment (devoncarew) [reply tool/lib/commands/packages_get.dart]: done

File: tool/lib/model.dart (hunk 9 above)
Comment (devoncarew) [reply tool/lib/model.dart]: done

We should update the analysis_options.yaml file at the root of the repo with the current style guide for this repo - it'll reduce code review churn.

File: tool/lib/model.dart (hunk 10 above)
Comment (devoncarew) [reply tool/lib/model.dart]: This remove the last segment of the path - `Platform.resolvedExecutable` returns the path to the Dart VM; this is removing the name of that binary.

**Comment by devoncarew** [comment]
I'm not excited about the CLI name (`devman`), but don't have anything better at the moment.

**Comment by devoncarew** [comment]
Hmm, perhaps `tools`.

**Comment by natebosch** [comment]
Can any of this be contributed to https://github.com/dart-lang/mono_repo instead of repeated locally?

**Comment by devoncarew** [comment]
> Can any of this be contributed to https://github.com/dart-lang/mono_repo instead of repeated locally?

Possibly, but that's likely out of scope for what I want to do here. Having this in source here will give us a lot more flexibility.

I'm not sure if mono_repo supports flutter packages, but if not that would be a useful thing to add to it.

**Pull request merged**